### ISSUE TITLE
Error importing recursive modules

### ISSUE THREAD

**Issue by TimurKiyivinski** [WebReflection/jsgtk#15]
First of all, thanks for this amazing library!

When trying it out, I ran into an error importing `ramda`. I believe this is related to #13 where I get the following error:
```
JS ERROR: Error: Circular dependencies detected:
.../node_modules/ramda/src/internal/_curry2
```
The following is minimal code that causes the error:
```javascript
#!/usr/bin/env jsgtk
const Gtk = require('Gtk');
const R = require('ramda');
```

Running it with `--debug` does not seem to give much information regarding `ramda`. I hope you have any ideas what might be going on, thanks in advance!

**Comment by WebReflection** [comment]
I'm afraid I won't have time to dig into their local dependencies graph hole:
https://github.com/ramda/ramda/blob/master/index.js

but it looks like they publish on npm the dist file too so writing the following worked for me:
```js
const R = require('ramda/dist/ramda.min.js');
```

**Comment by iddan** [comment]
This is not a real solution for the problem though 😕

**Comment by WebReflection** [comment]
PRs are welcome. Right now i don't have budget to fix all circular dependencies hell few libraries have.

If there is a workaround that works I'm already happy even if I perfectly know it's a workaround, not a solution.

There are so far only two libraries with this issue and both took circular dependencies to the next level.

I don't have resources to create another `npm` via old JS syntax supported in GJS so unless it's a real-world problem that cannot be solved otherwise I don't see myself wasting days behind this.

On top of that, GJS project is not really following me, this project, or all the hints or directions I've tried to give them. I'm waiting for them to finally release an npm compatible native functionality but it's not so transparent what's their timeline or how much resources they are putting there.

Instead of wasting my time branching out every day more from GJS, I'd rather wait to see where are they going for everyone maintenance sake.

**Comment by WebReflection** [comment]
P.S. to put in other words: this project, so far, is not sustainable and haven't received the contributions it was hoping for. I love all help I can receive for my open source projects but if it's a one off per year I cannot allocate much time to move forward. Thank you for your understanding.

**Comment by WebReflection** [comment]
Since OP was happy with the workaround, I'm also closing this as won't fix.

**Comment by iddan** [comment]
It's a shame Gnome doesn't give a damn about what's going in the JS community... Thank you so much for your time and contribution :)

**Comment by WebReflection** [comment]
they do but they are incredibly slow to move forward, they're based on Mozilla JS engine which is a pain to update each time, compared to latest V8, and they also assumed in the early days that having Python syntax in JS was a great idea.

What is a real shame is that node-gtk was the real-world fast moving-forward way to have all of this, but they never helped with the project which died.

JSGtk is the only used/useful survivor of all these attempts, yet it looks like GNOME doesn't even want Open Source and JS community help ... I don't know what to say.

**Comment by iddan** [comment]
@WebReflection is working for years on enhancing the GNOME-JS development experience. Sadly their seem to be no help or interest from the community / foundation with the subject.

Currently the JS community is one of the largest developers communities which has special interest in UI. Recent progress has changed the way web, mobile and desktop applications are being developed with projects like AngularJS, Ember, React, React Native and Electron.

I think it is time the GNOME community which has invested so much in JS will help us implement a better experience of developing GNOME apps with JS. 

@afranke @zeenix @cosimoc @allanday @MegFord @csoriano89

**Comment by TimurKiyivinski** [comment]
I think it's fair enough for a workaround, code doesn't come from nowhere and so far `ramda` is the only library with this issue for me~

**Comment by iddan** [comment]
For future users coming by: you can always use webpack as last resort

**Comment by zeenix** [comment]
@iddan The board doesn't steer the technical direction of the project. The contributors do that, like any real open source project out there. If you want GNOME to move into a particular direction, the best approach is to become a contributor and the more contributions you put in, the more power you gain to steer the ship. If you're interested: https://wiki.gnome.org/Newcomers

**Comment by WebReflection** [comment]
@iddan 

> you can always use webpack as last resort

no idea how you'd do that with GJS ... at runtime ... 

@zeenix 

> The board doesn't steer the technical direction of the project. The contributors do that

I've actively filed bugs and talks about better patterns for GJS in the ML, I feel like a contributor that tried to show, with a real Open Source project, what the JS community wanted.

As useless as a fact this is, JSGtk has more stars than GJS itself on GitHub ... why would I contribute to GJS if everything in there is the reason this project exists in the first place?

> like any real open source project out there

So you are saying that this is not a real open source project?

> the best approach is to become a contributor and the more contributions you put in, the more power you gain to steer the ship

I have many other **real** Open Source projects and when my users loudly need something I do steer my software to make them happy.

When alternative Open Source projects arrive with similar intents, even if 6 months late to the party, I understand what's the community need, ask them how to improve, and move to that direction.

You have an example [right here](https://github.com/WebReflection/hyperHTML/issues/79).

What GNOME did with GJS, was to tell JS developers they could write JS to create any Desktop app they wanted, and then it kinda forgot to listen to them, which is why mostly nobody knows it, or use it, and when they do, they realize this project is closer to what they need.

As summary, I think GNOME should slightly change its attitude if it wants to gain more developers adoption.

"_just contribute to gain more power in decisions_" doesn't seem the right approach, at least it didn't work so well so far, or did it?

**Comment by csoriano89** [comment]
Hey,

Adding to what Zeeshan say, there isn't a body which dictates where contributors have to work and also we do not employ any software developer.

If I understood correctly this project is an alternative of GJS, which sounds like a good thing to have. But I'm not sure what the issue is here and how we can help, do you mind elaborating given that we don't employ software engineers or that we don't mandate what contributors has to work on?

**Comment by WebReflection** [comment]
@csoriano89 

>  this project is an alternative of GJS

it's rather GJS on steroids, fully based on GJS too


> I'm not sure what the issue is here and how we can help

there's no issue _you_ can solve here, I think @iddan simply wanted to make you aware there is interest in this project and nobody from GNOME is helping at all.

To some extend, that'd be the only issue _we_ could maybe solve.


> we do not employ any software developer

who's _we_ and how is GNOME sustainable at all, if you don't mind me asking?

To clarify, I'm not looking for being employed at all, but I'm curious to know how the whole project moves forward in a healthy, future proof, way.

Thanks in advance for any sort of clarification.

**Comment by csoriano89** [comment]
>  I think @iddan simply wanted to make you aware there is interest in this project and nobody from GNOME is helping at all.
But GNOME is composed by single developers doing what they like to do, as in... maybe what you want is to talk with people interested in JS and convince them to contribute to your project? But that's the regular outreach any module inside GNOME does too.

> who's we and how is GNOME sustainable at all, if you don't mind me asking?

I think @iddan mentioned us because we are in the board of directors. You can check how the GNOME foundation works at https://www.gnome.org/foundation/

> To clarify, I'm not looking for being employed at all, but I'm curious to know how the whole project moves forward in a healthy, future proof, way.
Basically we share some vision, some people like it and they get involved to contribute to that vision. Usually they contribute to some existing project, then they become comfortable and experienced and eventually create projects on their own, and the project evolves its vision in this way. There is not much magic, to be honest :)

**Comment by iddan** [comment]
There is clearly *some* people in the GNOME community who decide which technologies to use and how to use them and clearly there was a decision working with JS. I'm just trying to understand how can we present the community a path towards a GNOME + JS environment that is coordinated with the JS community. I'm raising it here as @WebReflection is the only one I know who made progress towards this goal

**Comment by iddan** [comment]
I'm aware none of you is deciding but the community. But it is an important matter, for me at least, and as an outside force it's hard to reach the many awesome developers of the GNOME community. I'm only a user for about 8 years and a passionate developer who wants to help

**Comment by csoriano89** [comment]
@iddan 
Is basically those writing apps inside GNOME get together and most of them agree on using something and improve it, by themselves. If you were part of GNOME and wanted to use python, you would use python and improve python. That's why we have different languages inside GNOME :)

**Comment by iddan** [comment]
The reason I brought up your names is:
> determines what software is officially part of the Project,

**Comment by WebReflection** [comment]
> That's why we have different languages inside GNOME :)

AFAIK there are already 2 "official" JS projects there though ... GJS and CJS, I don't think a third project from scratch would be beneficial for anyone, I rather think GJS could steer direction in a non breaking way, yet moving forward.

Me contributing to GJS with the goal to start yet another JS bindings for GNOME as soon as I have "_more power_" looks, with all due respect, the most broken OSS path I could think of.

Such a waste of everyone time and resources. So no, it's not like that, I cannot start tomorrow a new GNOME project 'cause nobody would accept it officially right away.

Am I wrong?

**Comment by iddan** [comment]
As @WebReflection mentioned SpiderMonkey and missing contributions is kinda holding GJS back comparing to other existing JS environments such as Node.js. The descision to go with it was made I believe before Node.js becamed a community standard. Today using Node.js would be an obvious decision almost as supporting CPython. But I think it doesn't make sense GNOME will maintain it's  own runtime

**Comment by iddan** [comment]
@WebReflection @csoriano89 @zeenix there is a need from the community (or the foundation) to support a path towards better JS integration as making such project and then getting it rejected would be too expensive for everyone involved
